### PR TITLE
gumble softmax mindspore #14951

### LINKED ISSUE #21043: gumble_softmax
add mindspore gumble_softmax

- [gumble_softmax] #21042

### PR DESCRIPTION
**Pull request by sushmanthreddy**
[gumble_softmax] #21043 
close #21043

### CODE REVIEW

**Review by AryanSharma21 — COMMENTED** [review]
Hi @sushmanthreddy , please add close #issue_number in the PR description so that merge will automatically close the related issue . also can you give the screenshot of test running in local env

**Comment by sushmanthreddy** [comment]
> Hi @sushmanthreddy , please add close #issue_number in the PR description so that merge will automatically close the related issue . also can you give the screenshot of test running in local env

I have mentioned the issue number here above as 14951 ...the issue has multiple tasks
https://github.com/unifyai/ivy/issues/14951

**Comment by sushmanthreddy** [comment]
close #21043

**Comment by AryanSharma21** [comment]
well looking at the test log , I dont see any error regarding your implementation . You can request a review from maintainer for further check and merging

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong can u pls review the pr
it doest have any errors, till now pls review it and merge it

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong this implementation is for mindspore library , it has been mentioned in the to-do issue, still now we don't have support for mindspore but it can be implemented in future

**Comment by Sam-Armstrong** [comment]
@sushmanthreddy no, we need the function to be a pure composition of ivy functions. For example, this is the implementation of a similar function in the torch frontend https://github.com/unifyai/ivy/blob/75d8a7f32a935e1134b0be47ab8b044a534b4342/ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py#L336

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong can u pls review this once?

File: ivy/functional/frontends/mindspore/ops/function/nn_func.py [hunk 1]
```diff
@@ -36,3 +36,23 @@ def selu(input_x):
 @to_ivy_arrays_and_back
 def softsign(x):
     return ivy.divide(x, ivy.add(1, ivy.abs(x)))
+
+
+@with_supported_dtypes({"2.0.0 and below": "float16"}, "mindspore")
+@to_ivy_arrays_and_back
+def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/mindspore/ops/function/nn_func.py:43]: I think this function signature is wrong https://www.mindspore.cn/docs/zh-CN/r2.1/api_python/ops/mindspore.ops.gumbel_softmax.html?highlight=gumbel_softmax#mindspore.ops.gumbel_softmax. Looks like we don't need `eps`

File: ivy/functional/frontends/mindspore/ops/function/nn_func.py [hunk 2]
```diff
@@ -36,3 +36,23 @@ def selu(input_x):
 @to_ivy_arrays_and_back
 def softsign(x):
     return ivy.divide(x, ivy.add(1, ivy.abs(x)))
+
+
+@with_supported_dtypes({"2.0.0 and below": "float16"}, "mindspore")
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/mindspore/ops/function/nn_func.py:41]: `float32` is also supported

File: ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py [hunk 3]
```diff
@@ -77,3 +77,46 @@
...
+#     frontend,
+#     test_flags,
+# ):
+#     input_logits = logits
+#     gumbel_tau = tau
+#     gumbel_hard = hard
+#     gumbel_eps = eps
+#     gumbel_dim = dim
+
+#     helpers.test_frontend_function(
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py:112]: we also need to include `test_values=False` here, ala https://github.com/unifyai/ivy/blob/master/ivy_tests/test_ivy/test_frontends/test_torch/test_nn/test_functional/test_non_linear_activation_functions.py#L1398. also need to include `input_dtypes`. see the other tests in this file for reference, and make sure you're including everything required.

File: ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py [hunk 4]
```diff
@@ -77,3 +77,46 @@
...
+#         input_logits=input_logits,
+#         frontend=frontend,
+#         test_flags=test_flags,
+#         fn_tree=fn_tree,
+#         on_device=on_device,
+#         gumbel_tau=gumbel_tau,
+#         gumbel_hard=gumbel_hard,
+#         gumbel_eps=gumbel_eps,
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py:120]:
```suggestion
```

File: ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py [hunk 5]
```diff
@@ -77,3 +77,46 @@
...
+#     gumbel_eps = eps
+#     gumbel_dim = dim
+
+#     helpers.test_frontend_function(
+#         input_logits=input_logits,
+#         frontend=frontend,
+#         test_flags=test_flags,
+#         fn_tree=fn_tree,
+#         on_device=on_device,
+#         gumbel_tau=gumbel_tau,
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py:118]: these kwargs should be named the same as the function signature, so `tau=gumbel_tau` rather than `gumbel_tau=gumbel_tau` for example.

File: ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py [hunk 6]
```diff
@@ -77,3 +77,46 @@
...
+# @handle_frontend_test(
+#     fn_tree="mindspore.ops.function.nn_func.gumbel_softmax",
+#     logits=helpers.dtype_and_values(
+#         available_dtypes=helpers.get_dtypes("valid"),
+#         num_arrays=1,
+#         shared_dtype=True,
+#     ),
+#     tau=st.floats(min_value=0.1, max_value=10.0),
+#     hard=st.booleans(),
+#     eps=st.floats(min_value=1e-12, max_value=1e-6),
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py:91]: likewise, remove references to `eps` in the test

File: ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py [hunk 7]
```diff
@@ -77,3 +77,46 @@
...
+#     tau,
+#     hard,
+#     eps,
+#     dim,
+#     on_device,
+#     fn_tree,
+#     frontend,
+#     test_flags,
+# ):
+#     input_logits = logits
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_mindspore/test_ops/test_function/test_mindspore_nn_func.py:106]: this should be something like `input_dtype, logits = dtype_and_logits` instead of  `input_logits = logits`, as you're using `helpers.dtype_and_values`

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@sushmanthreddy I think this implementation should be fine assuming the function works exactly the same as torch's, which it looks like it does from the documentation. there are a few changes we need to make here though, particularly to the test - please can you look into these? thanks 🙂

**Comment by Sam-Armstrong** [comment]
also, please can you merge the master branch to resolve the merge conflicts 😁

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong can u pls review now?

i have removed merge conflicts and added the all changes requested.

File: ivy/functional/frontends/mindspore/ops/function/nn_func.py (hunk 1 above)
Comment (Sam-Armstrong) [reply ivy/functional/frontends/mindspore/ops/function/nn_func.py:43]: @sushmanthreddy we still need to remove `eps` here right?

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong i have made the changes can u pls review once??

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@sushmanthreddy please could you limit your changes to just the function and test you're implementing? Also I believe we still need to remove the `eps` arg. Thanks 😁

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong I have limited it to my code only, Those are of because of lint, the code is not in right format so, when I am running pre-commit those changes are firing up.we are not changing any code from,just we are setting up the format issues.

And due to merge conflicts these has also been reverted back to old format...But now eps value is removed..

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong I have made the changes,can u pls review it and merge it for one last time..if any changes are needed pls let me knew

**Review by Sam-Armstrong — APPROVED** [review]
thanks for your work on this @sushmanthreddy, lgtm!
